packages/muelu/adapters/xpetra/MueLu_RefMaxwell_def.hpp

```diff
@@ -500,110 +500,119 @@ namespace MueLu {
+            if (rebalanceStriding >= 1) {
+              bool acceptPart = (SM_Matrix_->getDomainMap()->getComm()->getRank() % rebalanceStriding) == 0;
+              if (SM_Matrix_->getDomainMap()->getComm()->getRank() >= numProcsAH*rebalanceStriding)
+                acceptPart = false;
+              repartParams.set("repartition: remap accept partition", acceptPart);
+            }
+            repartFactory->SetParameterList(repartParams);
+            // repartFactory->SetFactory("number of partitions", repartheurFactory);
+            repartFactory->SetFactory("Partition", partitioner);
+
+            auto newP = rcp(new RebalanceTransferFactory());
+            ParameterList newPparams;
+            newPparams.set("type", "Interpolation");
+            newPparams.set("repartition: rebalance P and R", precList11_.get<bool>("repartition: rebalance P and R", false));
+            newPparams.set("repartition: use subcommunicators", true);
+            newPparams.set("repartition: rebalance Nullspace", !NullspaceH_.is_null());
+            newP->SetFactory("Coordinates", NoFactory::getRCP());
+            if (!NullspaceH_.is_null())
+              newP->SetFactory("Nullspace", NoFactory::getRCP());
+            newP->SetParameterList(newPparams);
+            newP->SetFactory("Importer", repartFactory);
+
+            auto newA = rcp(new RebalanceAcFactory());
+            ParameterList rebAcParams;
+            rebAcParams.set("repartition: use subcommunicators", true);
+            newA->SetParameterList(rebAcParams);
+            newA->SetFactory("Importer", repartFactory);
+
+            coarseLevel.Request("P", newP.get());
+            coarseLevel.Request("Importer", repartFactory.get());
+            coarseLevel.Request("A", newA.get());
+            coarseLevel.Request("Coordinates", newP.get());
+            if (!NullspaceH_.is_null())
+              coarseLevel.Request("Nullspace", newP.get());
+            repartFactory->Build(coarseLevel);
+
+            if (!precList11_.get<bool>("repartition: rebalance P and R", false))
+              ImporterH_ = coarseLevel.Get< RCP<const Import> >("Importer", repartFactory.get());
+            P11_ = coarseLevel.Get< RCP<Matrix> >("P", newP.get());
+            AH_ = coarseLevel.Get< RCP<Matrix> >("A", newA.get());
+            CoordsH_ = coarseLevel.Get< RCP<RealValuedMultiVector> >("Coordinates", newP.get());
+            if (!NullspaceH_.is_null())
+              NullspaceH_ = coarseLevel.Get< RCP<MultiVector> >("Nullspace", newP.get());
+
+            AH_AP_reuse_data_ = Teuchos::null;
+            AH_RAP_reuse_data_ = Teuchos::null;
+
+            if (!disable_addon_ && enable_reuse_) {
+              // Rebalance the addon for next setup
+              RCP<const Import> ImporterH = coarseLevel.Get< RCP<const Import> >("Importer", repartFactory.get());
+              RCP<const Map> targetMap = ImporterH->getTargetMap();
+              SM_Matrix_->getDomainMap()->getComm()->barrier();
```

Comment:
Why is this necessary?

Reply:
We need to make sure that the addon term lives on the same set of MPI ranks as the coarse (1,1) problem. If we do not want to do reuse this isn't an issue, since the matrix for the coarse (1,1) problem will be set up on the global communicator before getting rebalanced to a sub-comm (with addon added it). If we do want reuse, then we have to update that matrix, but the addon part stays the same. We update the coarse matrix on the sub-communicator directly, but we then also need the addon there.